### ISSUE TITLE
`gather-with-grad` without `local-loss`?

### ISSUE THREAD

**Issue by zanussbaum** [mlfoundations/open_clip#616]
Hi all, thank you for the great repo!

I had a quick question regarding the loss function and args. Is `gather-with-grad` meant to be used only with `local-loss`? 

Comparing (and similar to https://github.com/pytorch/pytorch/issues/58005) it seems that using `torch.distributed.nn.all_gather` vs. `torch.distributed.all_gather` scales the gradient by the number of GPUs.

Here's an example

```python
import torch.nn as nn
import torch
import torch.distributed.nn
import torch.distributed as dist


def gather(tensor):
    gathered = [torch.empty_like(tensor) for _ in range(dist.get_world_size())]
    dist.all_gather(gathered, tensor)
    gathered[dist.get_rank()] = tensor 
    return torch.cat(gathered, dim=0)

def loss_comparision(use_distributed_gather=False):
    tensor = torch.ones(1, requires_grad=True).to(dist.get_rank()).unsqueeze(0)

    model = nn.Linear(1, 1, bias=False)
    model.weight.data[:] = 1.
    model = model.to(tensor.device)

    outputs = model(tensor) * 2
    if use_distributed_gather: 
        gathered_tensor = torch.cat(torch.distributed.nn.all_gather(outputs), dim=0)
    else:
        gathered_tensor = gather(outputs)

    loss = (gathered_tensor).sum()
    loss.backward()
    print(f"rank: {dist.get_rank()} loss with torch.distributed.all_gather: {loss}\n")
    print(f"rank: {dist.get_rank()} gradient with torch.distributed.all_gather: {model.weight.grad}\n")
    

if __name__ == "__main__":
    dist.init_process_group(backend='nccl')
    rank = dist.get_rank()
    torch.cuda.set_device(rank)
    loss_comparision(use_distributed_gather=False)
```

with  `use_distributed_gather=True`
```
rank: 0 loss with torch.distributed.all_gather: 4.0
rank: 1 loss with torch.distributed.all_gather: 4.0
rank: 0 gradient with torch.distributed.all_gather: tensor([[4.]], device='cuda:0')
rank: 1 gradient with torch.distributed.all_gather: tensor([[4.]], device='cuda:1')
```

and with False
```
rank: 0 loss with torch.distributed.all_gather: 4.0
rank: 1 loss with torch.distributed.all_gather: 4.0
rank: 0 gradient with torch.distributed.all_gather: tensor([[2.]], device='cuda:0')
rank: 1 gradient with torch.distributed.all_gather: tensor([[2.]], device='cuda:1')
```

**Comment by mitchellnw** [comment]
Hi Zach, yes `gather-with-grad` meant to be used only with `local-loss`, apologies that this is not more documented and thanks for raising the issue.

**Comment by zanussbaum** [comment]
No prob thanks for the quick reply @mitchellnw :D

**Comment by zanussbaum** [comment]
I had a follow up question and wanted to understand what's going on with `local-loss` and `gather-with-grad` a little better. 

Is there an intuitive explanation on what's going on? My naive reading of the code is that `local-loss` is calculating the CLIP loss for the batch image/text features per GPU by creating the `N/num_gpus X N` similarity matrix. Why is `gather-with-grad` needed here? Doesn't this scale the gradient by the number of GPUs? Is this to somehow simulate a "full batch" across all devices? 

Thank you :D

**Comment by Zasder3** [comment]
Good question! Your first observation is correct this reduces memory overhead by computing "slivers" of the traditional similarity matrix that are of shape `N/num_gpus x N`. So then why the `gather-with-grad`? This is needed because of the dot product computations we're doing. Let $a$ and $b$ be two vectors resulting from our text model and image model respectively. Differentiating their cosine similarity w.r.t. $\theta$ gives:

$$a\cdot b = (\nabla_\theta b) a+ (\nabla_\theta a) b$$

In particular let's say $a$ is on device and $b$ comes from the all gather. `torch.dist.all_gather` gives tensors without gradients making our value of $\nabla_\theta b=0$! That's a problem. On the other hand, `torch.dist.nn.all_gather` gives us tensors that will retain their computation graphs. So $\nabla_\theta b$ will give us what we were hoping for.

**Comment by zanussbaum** [comment]
oh hey cade funny seeing you here :) 

thanks for the explanation! where can i learn more about `torch.dist.all_gather` not having gradients? i didn't see anything in the docs/ in the code snippet above i ran it seems like `torch.dist.all_gather` still has gradients although they are scaled by the `1/ number of GPUs` present

**Comment by Zasder3** [comment]
haha this repo is the reason we met back at NeurIPS! :)

To see the lack of gradients first hand, you only need to adjust one line. Right before you calculate loss, try setting the on device computed value to 0. That is before `loss = (gathered_tensor).sum()` insert `gathered_tensor[dist.get_rank()] = 0`. After the changes you get the following:
```
With False
rank: 0 loss with torch.distributed.all_gather: 2.0
rank: 1 loss with torch.distributed.all_gather: 2.0

rank: 0 gradient with torch.distributed.all_gather: tensor([[0.]], device='cuda:6')
rank: 1 gradient with torch.distributed.all_gather: tensor([[0.]], device='cuda:7')

With True
rank: 0 loss with torch.distributed.all_gather: 2.0
rank: 1 loss with torch.distributed.all_gather: 2.0

rank: 0 gradient with torch.distributed.all_gather: tensor([[2.]], device='cuda:6')
rank: 1 gradient with torch.distributed.all_gather: tensor([[2.]], device='cuda:7')
```

You can see the origins of the engineering necessity in: https://github.com/openai/CLIP/issues/132. To get a feeling under the hood checkout the [backward pass](https://github.com/pytorch/pytorch/blob/2770e3addd9f05101705f0fef85a163e0034b8a5/torch/distributed/nn/functional.py#L329) of `nn.all_gather`. Compare that to the [vanilla](https://github.com/pytorch/pytorch/blob/2770e3addd9f05101705f0fef85a163e0034b8a5/torch/distributed/distributed_c10d.py#L2788) `all_gather`.

**Comment by zanussbaum** [comment]
ok so what happens in the case where `gather-with-grad` and `local-loss` is False? [Does this](https://github.com/mlfoundations/open_clip/blob/main/src/open_clip/loss.py#L56-L59) somehow preserver the gradient?

Thanks for answering all my dumb questions :)

**Comment by Zasder3** [comment]
More than happy to help these are good questions that not many people talk abt :)

The snippet you showed preserves gradients by splicing in a tensor you calculated with gradients into a tensor in which none of the elements do have a gradient. What you're inserting ends up becoming the only section of the all gathered tensor with gradients.

Overall, I'd say yes one should always do `gather-with-grad` and `local-loss` when doing a distributed training run. This will ensure a 1:1 matching of the loss/gradients to the non-sharded `NxN` matrix.

**Comment by zanussbaum** [comment]
thanks for the great explanations! this makes a ton of sense. closing
